From 1e2a2cd313c3580665e6003b41c7f44da126ef64 Mon Sep 17 00:00:00 2001
From: Peng Fan <peng.fan@nxp.com>
Date: Fri, 7 Sep 2018 16:23:16 +0800
Subject: [PATCH 4560/5242] MLK-19450-2 ARM64: dts: 8qm: support cm41 rpmsg
 with DomU

commit  e1291723ca2665195e30004b7dc75303524b5874 from
https://source.codeaurora.org/external/imx/linux-imx.git

1. Dom0 dts include fsl-imx8qm-mek.dtsi
2. Add /memreserve/ according to reserved-memory no-map node, then
   xen will not use these memory. The memory region are used by
   vpu/dsp/rpmsg, so xen should not touch them.
3. correct dom0 cma area, CM4 has limitation that the max access address
   is 0xE0000000, so the alloc-ranges should consider the limitation,
   otherwise rpmsg dma allocation will alloc memory higher than
   0xE0000000 and M4 will crash.
4. Hook CM41 with SMMU, added the addresses the CM41 will access, then
   after SMMU enabled, CM41 could access the address. To support
   Rear-View Camera, CM41 is kicked off by SCU at very early stage,
   DomU memory almost has no chance to have machine address 0x90000000
   included which is the vring desc buffer. So we have to enable SMMU
   to let CM41 access the memory.
5. Since DomU Guest RAM0 base is moved to 0x80000000, Let's change DomU
   ip address space to their machine address, since there is no conflict
   now.
6. Add reserved-memory in DomU dts, we enabled xen xl to copy that
   to DomU dtb.
7. Mark PCI/VPU as xen,passthrough, but not supported in DomU now.
8. Add Pixel_combiner2 passthrough to make dpu2 display work.

Signed-off-by: Peng Fan <peng.fan@nxp.com>
Acked-by: Anson Huang <Anson.Huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../boot/dts/freescale/fsl-imx8qm-mek-dom0.dts     |   95 ++++++++++++++-
 .../boot/dts/freescale/fsl-imx8qm-mek-domu.dts     |  127 ++++++++++++++------
 arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi  |   30 +++++
 3 files changed, 211 insertions(+), 41 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts
index d4ae5b3..db9f3f9 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts
@@ -12,10 +12,21 @@
  * GNU General Public License for more details.
  */
 
-#include "fsl-imx8qm-mek.dts"
+/dts-v1/;
+
+/memreserve/ 0x84000000 0x4000000;
+/memreserve/ 0x90000000 0x400000;
+/memreserve/ 0x90400000 0x2000000;
+/memreserve/ 0x92400000 0x2000000;
+/memreserve/ 0x94400000 0x1800000;
+
+#include "fsl-imx8qm-mek.dtsi"
 #include "fsl-imx8qm-xen.dtsi"
 
 / {
+	model = "Freescale i.MX8QM MEK DOM0";
+	compatible = "fsl,imx8qm-mek", "fsl,imx8qm";
+
 	chosen {
 		#address-cells = <2>;
 		#size-cells = <2>;
@@ -57,6 +68,7 @@
 				SC_R_MU_2A
 			>;
 			rsrcs = <
+				SC_R_MU_6A
 				SC_R_GPU_1_PID0
 				SC_R_GPU_1_PID1
 				SC_R_GPU_1_PID2
@@ -129,7 +141,7 @@
 			compatible = "shared-dma-pool";
 			reusable;
 			size = <0 0x28000000>;
-			alloc-ranges = <0 0xd0000000 0 0x28000000>;
+			alloc-ranges = <0 0xa0000000 0 0x40000000>;
 			linux,cma-default;
 		};
 	};
@@ -149,6 +161,46 @@
 		status = "disabled";
 	};
 
+	cm41: cm41@1 {
+		fsl,sc_rsrc_id = <SC_R_M4_1_PID0>,
+				 <SC_R_M4_1_PID1>,
+				 <SC_R_M4_1_PID2>,
+				 <SC_R_M4_1_PID3>,
+				 <SC_R_M4_1_PID4>;
+		#stream-id-cells = <1>;
+		iommus = <&smmu>;
+		xen,passthrough;
+	};
+
+	irqsteer_cm41: irqsteer_cm41@0x51080000 {
+		reg = <0x0 0x51080000 0x0 0x10000>;
+		xen,passthrough;
+	};
+
+	mu_rpmsg1_b: mu_rpmsg1_b@0x5d2a0000 {
+		reg = <0x0 0x5d2a0000 0x0 0x10000>;
+		xen,passthrough;
+	};
+};
+
+&mu_rpmsg1 {
+	xen,passthrough;
+};
+
+&rpmsg1 {
+	status = "disabled";
+};
+
+&mu_6_lpcg {
+	xen,passthrough;
+};
+
+&mu_6_lpcg_b {
+	xen,passthrough;
+};
+
+&mu_7_lpcg_b {
+	xen,passthrough;
 };
 
 &usbotg1_lpcg {
@@ -167,6 +219,19 @@
 	xen,passthrough;
 };
 
+/*
+ * DomU CM41 use this, but DomU OS not need this,
+ * because smmu is enabled for CM41, so need to
+ *create the lpuart2 mapping in SMMU
+ */
+&lpuart2 {
+	xen,passthrough;
+};
+
+&lpuart2_lpcg {
+	xen,passthrough;
+};
+
 &di_lvds1_lpcg {
 	xen,passthrough;
 };
@@ -186,7 +251,7 @@
 &smmu {
 	mmu-masters = <&dpu2 0x13>, <&gpu_3d1 0x15>,
 		      <&usdhc1 0x12>, <&usbotg1 0x11>,
-		      <&edma01 0x10>;
+		      <&edma01 0x10>, <&cm41 0x09>;
 };
 
 &lvds_region2 {
@@ -224,6 +289,10 @@
 	iommus = <&smmu>;
 };
 
+&pixel_combiner2 {
+	xen,passthrough;
+};
+
 &prg10 {
 	xen,passthrough;
 };
@@ -329,3 +398,23 @@
 	reg = <0x0 0x5b100000 0x0 0x1000>;
 	xen,passthrough;
 };
+
+&pciea {
+       xen,passthrough;
+};
+
+&pcieb {
+       xen,passthrough;
+};
+
+&dsp {
+       xen,passthrough;
+};
+
+&vpu_decoder {
+       xen,passthrough;
+};
+
+&vpu_encoder {
+       xen,passthrough;
+};
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
index 8bd72c6..03eba6a 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
@@ -66,7 +66,41 @@
 	memory@80000000 {
 		device_type = "memory";
 		/* Will be updated by U-Boot or XEN TOOL */
-		reg = <0x00000000 0x40000000 0 0x40000000>;
+		reg = <0x00000000 0x80000000 0 0x80000000>;
+	};
+
+	/*
+	 * The reserved memory will be used when using U-Boot loading android
+	 * image. For booting kernel using xl tool, pass args:
+	 * cma=960M@2400M-3584M
+	 * For the rpmsg_reserved area, need xl tool to create for non-android.
+	 */
+	reserved-memory {
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges;
+		passthrough;
+
+		/*
+		 * reserved-memory layout
+		 * 0x8800_0000 ~ 0x8FFF_FFFF is reserved for M4
+		 * Shouldn't be used at A core and Linux side.
+		 *
+		 */
+
+		rpmsg_reserved: rpmsg@0x90100000 {
+			no-map;
+			reg = <0 0x90100000 0 0x10000>;
+		};
+
+		/* global autoconfigured region for contiguous allocations */
+		linux,cma {
+			compatible = "shared-dma-pool";
+			reusable;
+			size = <0 0x3c000000>;
+			alloc-ranges = <0 0x96000000 0 0x3c000000>;
+			linux,cma-default;
+		};
 	};
 
 	gic: interrupt-controller@3001000 {
@@ -128,11 +162,10 @@
 			};
 		};
 
-
 		clk: clk {
 			compatible = "fsl,imx8qm-clk";
 			#clock-cells = <1>;
-			fsl,lpcg_base_offset = <0x00000001 0x00000000>;
+			fsl,lpcg_base_offset = <0x00000000 0x00000000>;
 		};
 
 		iomuxc: iomuxc {
@@ -141,22 +174,22 @@
 
 		#include "fsl-imx8qm-device.dtsi"
 
-		mu2: mu@15d1d0000 {
+		mu2: mu@5d1d0000 {
 			compatible = "fsl,imx8-mu";
-			reg = <0x1 0x5d1d0000 0x0 0x10000>;
+			reg = <0x0 0x5d1d0000 0x0 0x10000>;
 			interrupts = <GIC_SPI 178 IRQ_TYPE_LEVEL_HIGH>;
 			fsl,scu_ap_mu_id = <0>;
 			status = "okay";
 		};
 
 		usb_lpcg {
-			reg = <0x1 0x5b270000 0x0 0x10000>;
+			reg = <0x0 0x5b270000 0x0 0x10000>;
 		};
 
 		edma01: dma-controller1@5a1f0000 {
 			compatible = "fsl,imx8qm-edma";
-			reg = <0x1 0x5a2e0000 0x0 0x10000>, /* channel14 UART1 rx */
-			      <0x1 0x5a2f0000 0x0 0x10000>; /* channel15 UART1 tx */
+			reg = <0x0 0x5a2e0000 0x0 0x10000>, /* channel14 UART1 rx */
+			      <0x0 0x5a2f0000 0x0 0x10000>; /* channel15 UART1 tx */
 			#dma-cells = <3>;
 			dma-channels = <2>;
 			interrupts = <GIC_SPI 436 IRQ_TYPE_LEVEL_HIGH>,
@@ -176,87 +209,87 @@
 };
 
 &dpu2_intsteer {
-	reg = <0x1 0x57000000 0x0 0x10000>;
+	reg = <0x0 0x57000000 0x0 0x10000>;
 	status = "okay";
 };
 
 &prg10 {
-	reg = <0x1 0x57040000 0x0 0x10000>;
+	reg = <0x0 0x57040000 0x0 0x10000>;
 	status = "okay";
 };
 
 &prg11 {
-	reg = <0x1 0x57050000 0x0 0x10000>;
+	reg = <0x0 0x57050000 0x0 0x10000>;
 	status = "okay";
 };
 
 &prg12 {
-	reg = <0x1 0x57060000 0x0 0x10000>;
+	reg = <0x0 0x57060000 0x0 0x10000>;
 	status = "okay";
 };
 
 &prg13 {
-	reg = <0x1 0x57070000 0x0 0x10000>;
+	reg = <0x0 0x57070000 0x0 0x10000>;
 	status = "okay";
 };
 
 &prg14 {
-	reg = <0x1 0x57080000 0x0 0x10000>;
+	reg = <0x0 0x57080000 0x0 0x10000>;
 	status = "okay";
 };
 
 &prg15 {
-	reg = <0x1 0x57090000 0x0 0x10000>;
+	reg = <0x0 0x57090000 0x0 0x10000>;
 	status = "okay";
 };
 
 &prg16 {
-	reg = <0x1 0x570a0000 0x0 0x10000>;
+	reg = <0x0 0x570a0000 0x0 0x10000>;
 	status = "okay";
 };
 
 &prg17 {
-	reg = <0x1 0x570b0000 0x0 0x10000>;
+	reg = <0x0 0x570b0000 0x0 0x10000>;
 	status = "okay";
 };
 
 &prg18 {
-	reg = <0x1 0x570c0000 0x0 0x10000>;
+	reg = <0x0 0x570c0000 0x0 0x10000>;
 	status = "okay";
 };
 
 &dpr3_channel1 {
-	reg = <0x1 0x570d0000 0x0 0x10000>;
+	reg = <0x0 0x570d0000 0x0 0x10000>;
 	status = "okay";
 };
 
 &dpr3_channel2 {
-	reg = <0x1 0x570e0000 0x0 0x10000>;
+	reg = <0x0 0x570e0000 0x0 0x10000>;
 	status = "okay";
 };
 
 &dpr3_channel3 {
-	reg = <0x1 0x570f0000 0x0 0x10000>;
+	reg = <0x0 0x570f0000 0x0 0x10000>;
 	status = "okay";
 };
 
 &dpr4_channel1 {
-	reg = <0x1 0x57100000 0x0 0x10000>;
+	reg = <0x0 0x57100000 0x0 0x10000>;
 	status = "okay";
 };
 
 &dpr4_channel2 {
-	reg = <0x1 0x57110000 0x0 0x10000>;
+	reg = <0x0 0x57110000 0x0 0x10000>;
 	status = "okay";
 };
 
 &dpr4_channel3 {
-	reg = <0x1 0x57120000 0x0 0x10000>;
+	reg = <0x0 0x57120000 0x0 0x10000>;
 	status = "okay";
 };
 
 &dpu2 {
-	reg = <0x1 0x57180000 0x0 0x40000>;
+	reg = <0x0 0x57180000 0x0 0x40000>;
 	status = "okay";
 
 	dpu2_disp0: port@0 {
@@ -277,6 +310,10 @@
 	};
 };
 
+&pixel_combiner2 {
+	status = "okay";
+};
+
 /delete-node/ &hdmi;
 /delete-node/ &irqsteer_dsi0;
 /delete-node/ &i2c0_mipi_dsi0;
@@ -286,12 +323,12 @@
 /delete-node/ &mipi_dsi_bridge1;
 
 &lvds_region2 {
-	reg = <0x1 0x57240000 0x0 0x10000>;
+	reg = <0x0 0x57240000 0x0 0x10000>;
 	status = "okay";
 };
 
 &ldb2_phy {
-	reg = <0x1 0x57241000 0x0 0x100>;
+	reg = <0x0 0x57241000 0x0 0x100>;
 	status = "okay";
 };
 
@@ -331,6 +368,7 @@
 /delete-node/ &dpr2_channel2;
 /delete-node/ &dpr2_channel3;
 /delete-node/ &dpu1;
+/delete-node/ &pixel_combiner1;
 /delete-node/ &dsp;
 /delete-node/ &irqsteer_dsi1;
 /delete-node/ &i2c0_mipi_dsi1;
@@ -356,7 +394,7 @@
 /delete-node/ &i2c0_hdmi;
 
 &irqsteer_lvds1 {
-	reg = <0x1 0x57240000 0x0 0x1000>;
+	reg = <0x0 0x57240000 0x0 0x1000>;
 	/delete-property/ interrupt-parent;
 	status = "okay";
 };
@@ -366,7 +404,7 @@
 /delete-node/ &flexcan3;
 
 &i2c1_lvds1 {
-	reg = <0x1 0x57247000 0x0 0x1000>;
+	reg = <0x0 0x57247000 0x0 0x1000>;
 	status = "okay";
 };
 
@@ -381,7 +419,7 @@
 
 &lpuart1 {
 	/delete-property/ interrupt-parent;
-	reg = <0x1 0x5a070000 0 0x1000>;
+	reg = <0x0 0x5a070000 0 0x1000>;
 	dmas = <&edma01 15 0 0>, <&edma01 14 0 1>;
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_lpuart1>;
@@ -416,15 +454,15 @@
 /delete-node/ &pwm7;
 
 &gpu_3d1 {
-	reg = <0x0 0x24100000 0 0x40000>;
+	reg = <0x0 0x54100000 0 0x40000>;
 	status = "okay";
 };
 
 /delete-node/ &gpu_3d0;
 
 &imx8_gpu_ss {
-	/* xen guests have 3GB of low RAM @ 1GB */
-	reg = <0x0 0x40000000 0x0 0xc0000000>;
+	/* xen guests have 2GB of low RAM @ 2GB */
+	reg = <0x0 0x80000000 0x0 0x80000000>;
 	reg-names = "phys_baseaddr";
 	cores = <&gpu_3d1>;
 	status = "okay";
@@ -436,7 +474,7 @@
 	compatible = "fsl,imx8qm-usdhc", "fsl,imx6sl-usdhc";
 	/*interrupt-parent = <&gic>;*/
 	/delete-property/ interrupt-parent;
-	reg = <0x1 0x5b010000 0x0 0x10000>;
+	reg = <0x0 0x5b010000 0x0 0x10000>;
 };
 
 /delete-node/ &usdhc2;
@@ -445,13 +483,13 @@
 /delete-node/ &fec2;
 
 &usbmisc1 {
-	reg = <0x1 0x5b0d0200 0x0 0x200>;
+	reg = <0x0 0x5b0d0200 0x0 0x200>;
 };
 
 /delete-node/ &usbmisc2;
 
 &usbphy1 {
-	reg = <0x1 0x5b100000 0x0 0x200>;
+	reg = <0x0 0x5b100000 0x0 0x200>;
 };
 
 /delete-node/ &usbh1;
@@ -460,7 +498,7 @@
 /delete-node/ &usbphynop2;
 
 &usbotg1 {
-	reg = <0x1 0x5b0d0000 0x0 0x200>;
+	reg = <0x0 0x5b0d0000 0x0 0x200>;
 	/delete-property/ interrupt-parent;
 };
 
@@ -499,7 +537,20 @@
 
 /delete-node/ &intmux_cm40;
 /delete-node/ &intmux_cm41;
-/delete-node/ &imx_rpmsg;
+
+&rpmsg1{
+	/*
+	 * 64K for one rpmsg instance:
+	 */
+	vdev-nums = <1>;
+	reg = <0x0 0x90100000 0x0 0x10000>;
+	status = "okay";
+};
+
+&mu_rpmsg1 {
+	reg = <0x0 0x5d210000 0x0 0x10000>;
+};
+
 /delete-node/ &crypto;
 /delete-node/ &caam_sm;
 /delete-node/ &sc_pwrkey;
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi
index 984e40c..bd7bd52 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi
@@ -95,6 +95,11 @@
 		reg = <0x0 0x5a470000 0x0 0x10000>;
 	};
 
+	lpuart2_lpcg: lpcg@5a480000 {
+		compatible = "fsl,imx8qm-lpuart-lpcg";
+		reg = <0x0 0x5a480000 0x0 0x10000>;
+	};
+
 	di_lvds0_lpcg: lpcg@56243000 {
 		compatible = "fsl,imx8qm-di-lvds0-lpcg";
 		reg = <0x0 0x56243000 0x0 0x1000>;
@@ -114,6 +119,31 @@
 		compatible = "fsl,imx8qm-dc1-lpcg";
 		reg = <0x0 0x57010000 0x0 0x10000>;
 	};
+
+	mu_5_lpcg: lpcg@5d600000 {
+		compatible = "fsl,imx8qm-mu-lpcg";
+		reg = <0x0 0x5d600000 0x0 0x10000>;
+	};
+
+	mu_6_lpcg: lpcg@5d610000 {
+		compatible = "fsl,imx8qm-mu-lpcg";
+		reg = <0x0 0x5d610000 0x0 0x10000>;
+	};
+
+	mu_5_lpcg_b: lpcg@5d690000 {
+		compatible = "fsl,imx8qm-mu-lpcg";
+		reg = <0x0 0x5d690000 0x0 0x10000>;
+	};
+
+	mu_6_lpcg_b: lpcg@5d6a0000 {
+		compatible = "fsl,imx8qm-mu-lpcg";
+		reg = <0x0 0x5d6a0000 0x0 0x10000>;
+	};
+
+	mu_7_lpcg_b: lpcg@5d6b0000 {
+		compatible = "fsl,imx8qm-mu-lpcg";
+		reg = <0x0 0x5d6b0000 0x0 0x10000>;
+	};
 };
 
 /delete-node/ &edma0;
-- 
1.7.9.5

